Olá, Isabelle

Meu nome é Ramon. Fico feliz em revisar seu projeto. Ao ao longo do texto farei algumas observações sobre melhorias no código e também farei comentários sobre suas percepções sobre o assunto. Estarei aberto a feedbacks e discussões sobre o tema.

**Peço que mantenha e não altere os comentários que eu fizer por aqui para que possamos nos localizar posteriormente, ok?**

Mais uma coisa, vamos utilizar um código de cores para você entender os meus feedbacks no seu notebook. Funciona assim:

<div class="alert alert-block alert-success">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Sucesso. Tudo foi feito corretamente.
</div>

<div class="alert alert-block alert-warning">
<b>Comentário do revisor: </b> <a class="tocSkip"></a>

Alerta não crítco, mas que pode ser corrigido para melhoria geral no seu código/análise
</div>

<div class="alert alert-block alert-danger">

<b>Comentário do revisor: </b> <a class="tocSkip"></a>
    
Erro que precisa ser consertado, caso contrário seu projeto **não** será aceito.
</div>

Você pode me responder usando isso:
<div class="alert alert-block alert-info">
<b>Resposta do Aluno.</b> <a class="tocSkip"></a>
</div>

1. **Analisar a popularidade e a recepção dos livros**: Este estudo pode visar entender quais livros foram mais bem recebidos pelos usuários, medindo a média das classificações e o número total de avaliações.

2. **Identificar tendências de publicação por editora**: O estudo pode incluir uma análise das editoras que lançaram mais livros com mais de 50 páginas, visando entender as tendências de publicação e a produção editorial no mercado.

3. **Avaliar a influência dos autores na classificação dos livros**: O objetivo pode ser identificar os autores cujos livros têm a média mais alta de classificação, considerando apenas os livros com um número significativo de avaliações (por exemplo, mais de 50).

4. **Entender o comportamento dos usuários na plataforma**: O estudo pode visar compreender o comportamento dos usuários que classificaram mais de 50 livros, analisando a média de avaliações atribuídas por esses usuários.

5. **Fornecer insights para decisões editoriais e de marketing**: Com base nos resultados das análises, os insights podem ser usados para tomar decisões relacionadas à publicação de novos livros, promoção de autores ou estratégias de marketing.

In [1]:
# importar bibliotecas
import pandas as pd
from sqlalchemy import create_engine
db_config = {'user': 'practicum_student', # nome de usuário
'pwd': 's65BlTKV3faNIGhmvJVzOqhs', # senha
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # porta de conexão
'db': 'data-analyst-final-project-db'} # o nome do banco de dados
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
db_config['pwd'],
db_config['host'],
db_config['port'],
db_config['db'])
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

In [2]:
query = """
    SELECT b.book_id,
           COUNT(r.rating_id) AS num_reviews,
           AVG(r.rating) AS avg_rating
    FROM books b
    LEFT JOIN ratings r ON b.book_id = r.book_id
    GROUP BY b.book_id
"""

df = pd.read_sql_query(query, engine)

print(df)

     book_id  num_reviews  avg_rating
0        652            2    4.500000
1        273            2    4.500000
2         51           12    4.250000
3        951            2    4.000000
4        839            7    4.285714
..       ...          ...         ...
995       64           13    4.230769
996       55            2    5.000000
997      148            7    3.428571
998      790            2    3.500000
999      828            2    3.000000

[1000 rows x 3 columns]


- `book_id`: Identificador único de cada livro.
- `num_reviews`: Número total de avaliações para cada livro.
- `avg_rating`: Classificação média atribuída a cada livro com base nas avaliações.

In [3]:
query = """
    SELECT COUNT(*)
    FROM books
    WHERE publication_date > '2000-01-01'
"""

df = pd.read_sql_query(query, engine)

print("Número de livros lançados depois de 1 de janeiro de 2000:", df.iloc[0, 0])

Número de livros lançados depois de 1 de janeiro de 2000: 819


In [4]:
query = """
    SELECT b.book_id,
           COUNT(r.rating_id) AS num_reviews,
           AVG(r.rating) AS avg_rating
    FROM books b
    LEFT JOIN ratings r ON b.book_id = r.book_id
    GROUP BY b.book_id
"""

df = pd.read_sql_query(query, engine)

print(df)

     book_id  num_reviews  avg_rating
0        652            2    4.500000
1        273            2    4.500000
2         51           12    4.250000
3        951            2    4.000000
4        839            7    4.285714
..       ...          ...         ...
995       64           13    4.230769
996       55            2    5.000000
997      148            7    3.428571
998      790            2    3.500000
999      828            2    3.000000

[1000 rows x 3 columns]


Estes resultados mostram informações sobre os livros, incluindo o `book_id`, o número de avaliações (`num_reviews`) e a classificação média (`avg_rating`). Cada linha representa um livro e suas respectivas estatísticas de avaliação.

In [5]:
query = """
    SELECT p.publisher_id, 
           p.publisher, 
           COUNT(b.book_id) AS num_books
    FROM publishers p
    JOIN books b ON p.publisher_id = b.publisher_id
    WHERE b.num_pages > 50
    GROUP BY p.publisher_id, p.publisher
    ORDER BY COUNT(b.book_id) DESC
    LIMIT 1
"""

df = pd.read_sql_query(query, engine)

print(df)

   publisher_id      publisher  num_books
0           212  Penguin Books         42


Estes resultados mostram que a editora com o ID 212, chamada "Penguin Books", lançou um total de 42 livros com mais de 50 páginas. 

In [6]:
query = """
    SELECT a.author_id, 
           a.author, 
           AVG(r.rating) AS avg_rating
    FROM authors a
    JOIN books b ON a.author_id = b.author_id
    JOIN ratings r ON b.book_id = r.book_id
    GROUP BY a.author_id, a.author
    HAVING COUNT(r.rating_id) > 50
    ORDER BY AVG(r.rating) DESC
    LIMIT 1
"""

df = pd.read_sql_query(query, engine)

print(df)

   author_id                      author  avg_rating
0        236  J.K. Rowling/Mary GrandPré    4.288462


O autor com o ID 236, que é J.K. Rowling/Mary GrandPré, com uma média de classificação de aproximadamente 4.29. Isso corresponde ao autor com a média mais alta de classificação de livros, considerando apenas os autores com mais de 50 avaliações.

In [7]:
query = """
SELECT username, COUNT(*) AS num_reviews
FROM reviews
GROUP BY username;
"""

df = pd.read_sql_query(query, engine)

print("Número de revisões por usuário:")
print(df)

Número de revisões por usuário:
           username  num_reviews
0         zjohnston           14
1           jnelson           21
2      lewisdesiree           19
3     patrickhudson           15
4             ulowe           16
..              ...          ...
155  shermannatalie           24
156         dmiller           18
157  charlesraymond           16
158     isaiahreyes           15
159       masonsara           14

[160 rows x 2 columns]


Esses resultados mostram o número de revisões feitas por cada usuário na plataforma. Cada linha corresponde a um usuário e ao número de revisões que eles fizeram.

In [8]:
query_total_reviews = """
SELECT COUNT(reviews.text)
FROM (
    SELECT username
    FROM ratings
    GROUP BY username
    HAVING COUNT(rating) > 50
) AS users
JOIN reviews ON users.username = reviews.username
GROUP BY users.username
"""

df_total_reviews = pd.read_sql_query(query_total_reviews, engine)

print("\nNúmero total de revisões feitas pelos usuários que classificaram mais de 50 livros:")
print(df_total_reviews)


Número total de revisões feitas pelos usuários que classificaram mais de 50 livros:
   count
0     22
1     18
2     27
3     26
4     28
5     25


In [9]:
avg_total_reviews = df_total_reviews.mean()

print("\nMédia do número total de revisões feitas pelos usuários que classificaram mais de 50 livros:")
print(avg_total_reviews)


Média do número total de revisões feitas pelos usuários que classificaram mais de 50 livros:
count    24.333333
dtype: float64


<div class="alert alert-block alert-warning">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Correto. O ideal era obter a média já usando SQL. Mas sem problemas.
</div>

Este resultado indica que, em média, os usuários que classificaram mais de 50 livros têm aproximadamente 24.33 avaliações. Isso fornece uma visão do comportamento dos usuários que são bastante ativos na plataforma de classificação de livros. 

<div class="alert alert-block alert-danger">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Esta seção diverge com o resultado do gabarito. Veja o código acima como exemplo. A diferença está no >= que você usou
</div>

In [10]:
query = '''
SELECT authors.author, AVG(ratings.rating), COUNT(ratings.book_id)
FROM authors
JOIN books ON authors.author_id = books.author_id
JOIN ratings ON books.book_id = ratings.book_id
GROUP BY authors.author
HAVING COUNT(ratings.book_id) > 50
ORDER BY AVG(ratings.rating) DESC
LIMIT 1
'''# Executando a consulta SQL e armazenando os resultados em um DataFrame


df = pd.read_sql_query(query, engine)

# Exibindo o resultado
print(df)

                       author       avg  count
0  J.K. Rowling/Mary GrandPré  4.288462    312


<div class="alert alert-block alert-danger">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

A resposta está incorreta. Precisamos primeiro contar quantas reviews por usuário. Veja este código abaixo. A resposta correta é 24.33
</div>

<div class="alert alert-block alert-info">
<b>Resposta do Aluno.</b> <a class="tocSkip"></a> Fiz as correções necessárias para atender aos resultados esperados.
</div>

<div class="alert alert-block alert-success">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Correto
</div>

In [11]:
query = """

     SELECT COUNT(reviews.text)
    FROM(
        SELECT username
        FROM ratings
        GROUP BY username
        HAVING COUNT(rating) > 50) AS users
    JOIN reviews ON users.username = reviews.username
    GROUP BY users.username
"""

df = pd.read_sql_query(query, engine)

print(df)

   count
0     22
1     18
2     27
3     26
4     28
5     25


1. **Número de livros lançados depois de 1 de janeiro de 2000:** Com base na contagem realizada, foram lançados 819 livros após 1 de janeiro de 2000. Isso pode indicar uma tendência de produção editorial mais recente ou uma expansão do mercado de livros após essa data.

2. **Número de avaliações e classificação média para cada livro:** Os dados mostram informações sobre os livros, incluindo o número de avaliações e a classificação média atribuída pelos usuários. Essas estatísticas podem ser úteis para avaliar a popularidade e a recepção de cada livro na plataforma.

3. **Editora que lançou o maior número de livros com mais de 50 páginas:** Com base na contagem realizada, a editora "Penguin Books" lançou o maior número de livros com mais de 50 páginas, totalizando 42 livros. Isso pode indicar uma estratégia editorial focada em publicar livros mais substanciais.

4. **Autor com a média mais alta de classificação de livros (considerando apenas livros com pelo menos 50 classificações):** O autor Diana Gabaldon tem a média mais alta de classificação de livros, com uma média de 4.29. Isso sugere que os livros de J.K. Rowling/Mary GrandPré são bem recebidos pelos leitores, especialmente considerando que a análise considerou apenas livros com pelo menos 50 classificações.

5. **Número médio de avaliações entre usuários que classificaram mais do que 50 livros:** Em média, os usuários que classificaram mais de 50 livros têm aproximadamente 24.33 avaliações. Isso fornece uma visão do comportamento dos usuários mais ativos na plataforma de classificação de livros.


<div class="alert alert-block alert-success">
<b>Comentário Geral do Revisor</b> <a class="tocSkip"></a>

Obrigado por enviar seu projeto. 
    
Parabéns, você fez um bom trabalho.

Como pontos fortes do projeto, destaco:
- SQL simples
- Uso adequado dos recursos da lição

    
<span class="text-danger">Porém, há espaço para melhorias conforme os itens apontados anteriormente. Destaco os  pontos a seguir: </span>
 * <span class="text-danger"> Há dois pequenos erros. Deixei dicas</span>


Portanto, sugiro que você corrija os pontos indicados.
    
Se você tiver alguma dúvida sobre de algo que você não tenha entendido, sinta-se à vontade para perguntar ao tutor ou deixar um comentário em **azul** para que, na próxima iteração, eu possa responder.

Conte conosco. Lembre-se de que você pode tirar dúvidas com revisores usando os comentários ou buscar ajuda de um tutor.
Bom trabalho.
</div>


<div class="alert alert-block alert-success">
<b>Comentário Geral do Revisor</b> <a class="tocSkip"></a>

Tudo certo agora. Parabéns pela aprovação.
</div>